### ISSUE TITLE
Deadlock stops all concourse containers

### ISSUE THREAD

**Issue by samrees** [concourse/concourse#5947]
## Summary

We had networking errors in our underlying Kubernetes cluster on AWS, seemingly routes between nodes, both in and out just stopped for a handful of seconds.

Out of that, everything had issues, but recovered. 

Later, we all concourse jobs stop, Lidar checks go to zero, and the postgresql DB reports this deadlock at that point:

```
2020-08-04 17:49:02 UTC:10.41.56.175(53938):concourse@concourse:[26687]:ERROR: deadlock detected
2020-08-04 17:49:02 UTC:10.41.56.175(53938):concourse@concourse:[26687]:DETAIL: Process 26687 waits for ShareLock on transaction 602883317; blocked by process 26449.
Process 26449 waits for ShareLock on transaction 602883320; blocked by process 26687.
Process 26687: DELETE FROM resource_config_check_sessions WHERE expires_at < NOW()
Process 26449: DELETE FROM workers WHERE ephemeral = $1 AND expires < NOW() RETURNING name
2020-08-04 17:49:02 UTC:10.41.56.175(53938):concourse@concourse:[26687]:HINT: See server log for query details.
2020-08-04 17:49:02 UTC:10.41.56.175(53938):concourse@concourse:[26687]:CONTEXT: while updating tuple (15,16) in relation "containers"
SQL statement "UPDATE ONLY "public"."containers" SET "resource_config_check_session_id" = NULL WHERE $1 OPERATOR(pg_catalog.=) "resource_config_check_session_id""
2020-08-04 17:49:02 UTC:10.41.56.175(53938):concourse@concourse:[26687]:STATEMENT: DELETE FROM resource_config_check_sessions WHERE expires_at < NOW()
```

* Concourse version: 6.3

This may be difficult to reproduce without the busted networking.

**Comment by risicle** [comment]
I *think* we experienced something very similar to this on Concourse 7.2.0 recently. (This is *separate* to the issue we documented in #7316)

Again, a concourse-web instance running out of memory began our incident, after which we observed long-running instances of the following 4 interrelated queries in RDS performance insights:

```
INSERT INTO workers (expires,start_time,addr,active_containers,active_volumes,resource_types,tags,platform,baggageclaim_url,certs_path,http_proxy_url,https_proxy_url,no_proxy,name,version,state,team_id,ephemeral) VALUES (NOW() + ?::INTERVAL,to_timestamp(?),$1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16) 
            ON CONFLICT (name) DO UPDATE SET
                expires = NOW() + ?::INTERVAL,
                start_time = to_timestamp(?),
                addr = $17,
                active_containers = $18,
                active_volumes = $19,
                resour...
```

```
UPDATE containers SET missing_since = now() WHERE (handle IN ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,$17,$18,$19) AND state <> $20)
```

```
DELETE FROM workers WHERE ephemeral = $1 AND expires < NOW() RETURNING name
```

```
DELETE FROM resource_config_check_sessions WHERE expires_at < NOW()
```

![image](https://user-images.githubusercontent.com/807447/126505583-ca52acb6-443d-41e1-b5ac-4b507fea7c8a.png)
(note the exact 1, 1 & 2, suggesting sessions spent the entire measured minute period executing that query)

After 6 minutes of this, postgres finally detects a deadlock and kills several of these queries (still a mystery why it took 6 minutes).

For the next hour or so no jobs get scheduled until the oom killer steps in and the out-of-memory concourse-web process gets restarted (on 10.1.10.78 in this plot)

![image](https://user-images.githubusercontent.com/807447/126506637-9cfcebbe-b330-4aed-bb16-f2201304c15b.png)

My main take-away from this experience was: don't let concourse-web run out of memory because it can cause weird things to happen.

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
